### PR TITLE
Add dotnet restore error recording

### PR DESCRIPTION
**Pull request by wesrupert**
- Add error output parsing to dotnet restore and dotnet publish so that we get more information than "Error: 1"
- Add basic platform data to telemetry properties (dotnet version, OS platform/runtime id)

@gregg-miskelly @caslan @rajkumar42 @chuckries

### CODE REVIEW

**Comment by wesrupert** [comment]
Not sure why the cla-required tag is showing up again...

**Comment by gregg-miskelly** [comment]
I manually fixed the tags...

File: src/coreclr-debug/activate.ts [hunk 1]
```diff
@@ -40,15 +40,33 @@ export function activate(context: vscode.ExtensionContext, reporter: TelemetryRe
         return;
     }
 
+    let dotnetVersion: string = '';
+    let osVersion: string = '';
+    let osRID: string = '';
+    _util.spawnChildProcess('dotnet', ['--info'], _util.coreClrDebugDir(), (data: Buffer) => {
```
Comment (gregg-miskelly) [inline src/coreclr-debug/activate.ts]: >_util [](start = 4, length = 5)

The call on line 30 is just a call to 'dontet --version' to see if it dotnet can successfully launch. We should probably combine these calls together so we don't have the extra spawn for no reason.

Comment (gregg-miskelly) [inline src/coreclr-debug/activate.ts]: >_util [](start = 4, length = 5)

This returns a promise that we need to wait on

File: src/coreclr-debug/util.ts [hunk 2]
```diff
@@ -109,7 +110,34 @@ export class CoreClrDebugUtil
...
+            child.stderr.on('data', onStderr);
+            child.on('close', (code: number) => {
+                if (code != 0) {
+                    this.log(`${process} exited with error code ${code}`);;
+                    reject(new Error(code.toString()));
+                }
+                else {
+                    resolve();
+                }
+            });
```
Comment (gregg-miskelly) [inline src/coreclr-debug/util.ts]: >}); [](start = 12, length = 3)

You may need an 'error' handler too -- from [https://nodejs.org/api/child_process.html#child_process_child_process_spawn_command_args_options](https://nodejs.org/api/child_process.html#child_process_child_process_spawn_command_args_options) see 'Example of checking for failed exec:'

File: src/coreclr-debug/install.ts [hunk 3]
```diff
@@ -12,46 +12,92 @@ import * as fs_extra from 'fs-extra-promise';
...
+                (data: Buffer) => {
+                    var text: string = data.toString();
+                    this._util.log(text);
+
+                    // Certain errors are only logged to stdout. Detect these and make a note of the kind of error.
+                    DebugInstaller.parseRestoreErrors(text, errorBuilder);
+                },
+                (data: Buffer) => {
+                    // Reference errors are sent to stderr at the end of restore.
+                    DebugInstaller.parseReferenceErrors(data.toString(), errorBuilder);
```
Comment (gregg-miskelly) [inline src/coreclr-debug/install.ts]: >data.toString() [](start = 56, length = 15)

Shouldn't we log this?

**Comment by gregg-miskelly** [comment]
Otherwise LGTM

**Comment by chuckries** [comment]
:+1:

File: src/coreclr-debug/activate.ts [hunk 4]
```diff
@@ -21,62 +21,84 @@ export function activate(context: vscode.ExtensionContext, reporter: TelemetryRe
...
+        let moreErrors: string = '';
+
+        writeInstallBeginFile().then(() => {
+            return installer.install(runtimeId);
+        }).then(() => {
+            installStage = "completeSuccess";
+            statusBarMessage.dispose();
+            vscode.window.setStatusBarMessage('Successfully installed .NET Core Debugger.');
+        })
+            .catch((error: debugInstall.InstallError) => {
```
Comment (gregg-miskelly) [inline src/coreclr-debug/activate.ts]: >catch [](start = 13, length = 5)

The indentation of this block seems weird. Shouldn't this be 4 spaces to the left?

**Comment by gregg-miskelly** [comment]
Otherwise LGTM

File: src/coreclr-debug/activate.ts (hunk 4 above)
Comment (wesrupert) [inline src/coreclr-debug/activate.ts]: It should be part of the line above, not sure what happened; reverted.

**Pull request merged**